Branch for all progress

MihailNenahov opened the pull request.

Commit:
added controller and application class and checked is it working

added city class

Commit:
Create project model structure

Removed City class

arramanenka commented on pom.xml:
```diff
@@ -8,31 +8,45 @@
     <artifactId>LogisticTask</artifactId>
     <version>1.0-SNAPSHOT</version>
 
-    <parent>
```
why?

MihailNenahov replied:
outdated

arramanenka commented on src/main/java/com/logistic/project/model/Warehouse.java:
```diff
@@ -1,4 +1,16 @@
+    private String name;
+    private Map<Resource, Integer> resourceQuantity;
+
+    public Warehouse(String name, Map<Resource, Integer> resourceQuantity) {
```
`@RequiredArgsConstructor`

MihailNenahov replied:
@Data is a convenient shortcut annotation that bundles the features of @ToString, @EqualsAndHashCode, @Getter / @Setter and @RequiredArgsConstructor together: In other words, @Data generates all the boilerplate that is normally associated with simple POJOs (Plain Old Java Objects) and beans: getters for all fields, setters for all non-final fields, and appropriate toString, equals and hashCode implementations that involve the fields of the class, and a constructor that initializes all final fields, as well as all non-final fields with no initializer that have been marked with @NonNull, in order to ensure the field is never null.

arramanenka commented on src/main/java/com/logistic/project/model/Route.java:
```diff
@@ -0,0 +1,4 @@
+package com.logistic.project.model;
+
+public class Route {
```
?

arramanenka commented on src/main/java/com/logistic/project/model/Resource.java:
```diff
@@ -0,0 +1,7 @@
+package com.logistic.project.model;
+
+public enum Resource {
```
What for?

arramanenka commented on src/main/java/com/logistic/project/controller/LogisticController.java:
```diff
@@ -0,0 +1,12 @@
+@RestController
+public class LogisticController {
+    @GetMapping("/greet")
+    public String greet() {
```
Remove it before merge

arramanenka commented on src/main/java/com/logistic/project/model/Order.java:
```diff
@@ -1,4 +1,10 @@
 public class Order {
+    private Customer to;
+    private Warehouse from;
+    private int distance;
```
distance is not really needed, is it?

arramanenka replied:
Also, when an order is placed, it is an order for an item (resource in your system), but Order in here is just from to info

Commit:
Added application.properties file

Added repositories
Added ids for model classes

Commit:
Made Resourses String

Commit:
renamed Resource to Merchandise

Commit:
Added configuration file with cost data

Added repository methods
Added util classes
Changed distance calculating logic
Added new DTOs

Commit:
Refactored Warehouse repository

Refactored MakeOrderDTO
Finished LogisticController

arramanenka commented on src/main/java/com/logistic/project/controller/LogisticController.java:
```diff
@@ -0,0 +1,80 @@
+    private final WarehouseRepository warehouseRepository;
+
+    @Autowired
+    public LogisticController(CustomerRepository customerRepository, WarehouseRepository warehouseRepository) {
```
`@RequiredArgsConstructor`

arramanenka commented on src/main/java/com/logistic/project/controller/LogisticController.java:
```diff
@@ -0,0 +1,80 @@
+    @PostMapping
+    public String createNewCustomer(Customer customer) {
+        customerRepository.save(customer);
+        return "response:200";
```
return customer instead of string

arramanenka commented on src/main/java/com/logistic/project/controller/LogisticController.java:
```diff
@@ -0,0 +1,80 @@
+    public Order getOrderRouteAndCost(MakeOrderDTO makeOrderDTO) {
+        Customer customer = customerRepository.findCustomerByName(makeOrderDTO.getCustomerDTO().getName());
+
+        Order finalOrder = new Order(null, customer, makeOrderDTO.getMerchandiseQuantity(), null);
```
builder

arramanenka commented on src/main/java/com/logistic/project/controller/LogisticController.java:
```diff
@@ -0,0 +1,80 @@
+
+        double cost = 0;
+        for (Warehouse warehouse : finalOrder.getWarehouses()) {
+            double merchandiseDeliveryCost = CoordinateCalculator.getCoordinateDistance(customer.getPosition(), warehouse.getPosition()) *
```
Cost is not in task

arramanenka commented on src/main/java/com/logistic/project/model/Order.java:
```diff
@@ -1,4 +1,31 @@
+
+@Data
 public class Order {
+    private Set<Warehouse> warehouses;
```
`=new HashSet<>()`

MihailNenahov replied:
Idea shows warning that it's redundant code

arramanenka commented on src/main/java/com/logistic/project/model/Warehouse.java:
```diff
@@ -1,4 +1,18 @@
 public class Warehouse {
+    private String name;
+    private Map<String, Integer> merchandiseQuantity;
+    private Coordinate position;
```
?

MihailNenahov replied:
outdated

arramanenka commented on src/main/java/com/logistic/project/util/Configuration.java:
```diff
@@ -0,0 +1,13 @@
+@PropertySource("classpath:LogicConfiguration.properties")
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class Configuration {
+    @Value("${mile.cost}")
```
Not needed

arramanenka commented on src/main/java/com/logistic/project/util/WarehouseCustomerDistanceComparator.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public int compare(Warehouse o1, Warehouse o2) {
+        int result = 0;
+
```
Comparator.comparingDouble( result of o1, result of o2)

Commit:
Fixed comments

arramanenka commented on src/main/java/com/logistic/project/controller/LogisticController.java:
```diff
@@ -0,0 +1,56 @@
+                merchandiseQuantity(makeOrderDTO.getMerchandiseQuantity()).
+                build();
+
+        for (String merchandiseName : makeOrderDTO.getMerchandiseQuantity().keySet()) {
```
comment out or delete before merging

Commit:
Finished writing repository layer

Some changes for model
Added to Order distance field
Added Mongo Configuration
Finished logic implementation

arramanenka commented on src/main/java/com/logistic/project/configuration/MongoConfig.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Bean
+    public MongoTemplate mongoTemplate() throws Exception {
+        return new MongoTemplate(mongo(), "test");
```
Instead of directly calling a method, just add mongoclient to arguments of the method. It will be autowired automatically

arramanenka commented on src/main/java/com/logistic/project/configuration/MongoConfig.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.data.mongodb.core.MongoTemplate;
+
+@Configuration
+public class MongoConfig {
```
If you use spring boot, which you do, you don't need this custom config. Spring boot automatically declares mpngotemplate, provided that there are necessary parameters in application.properties

Commit:
Added endpoints names

Commit:
Removed redundant builder set
